### PR TITLE
add phase_space func for gaussian state

### PR DESCRIPTION
**Pull request by sylviemonet**
**Context:** Add the function phase_space for state to be able to get the s-paramaterized phase space information of any state.

**Description of the Change:** 

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: mrmustard/lab_dev/circuit_components_utils.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+from typing import Sequence
+
+from mrmustard.physics import triples
+from .circuit_components import CircuitComponent
+from ..physics.representations import Bargmann
+
+__all__ = ["DsMap", "CftMap"]
+
+
+class DsMap(CircuitComponent):
```
Comment (sylviemonet) [inline mrmustard/lab_dev/circuit_components_utils.py:33]:
```suggestion
class _DsMap(CircuitComponent):
```

File: .github/CHANGELOG.md [hunk 2]
```diff
@@ -1,12 +1,19 @@
 # Current develop
 
 ### New features
-  
+* Added two new internal CircuitComponent ``DsMap`` and `CftMap` to convert the representations.
```
Comment (SamFerracin) [inline .github/CHANGELOG.md:4]: This is all happenening in lab_dev, and we don't put it in the changelog. Just mention the fact that you added a triple

File: mrmustard/lab_dev/circuit_components_utils.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    r"""The s-parametrized Dgate as a channel.
```
Comment (SamFerracin) [inline mrmustard/lab_dev/circuit_components_utils.py:34]:
```suggestion
    r"""The `s`-parametrized ``Dgate`` as a ``Channel``.
```

File: mrmustard/lab_dev/circuit_components_utils.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    This will be used as an internal Channel for representation transformation.
```
Comment (SamFerracin) [inline mrmustard/lab_dev/circuit_components_utils.py:35]:
```suggestion
    
    Used internally as a ``Channel`` for transformations between representations.
```

File: mrmustard/lab_dev/circuit_components_utils.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+    Args:
+        num_modes: number of modes of this channel.
```
Comment (SamFerracin) [inline mrmustard/lab_dev/circuit_components_utils.py:38]:
```suggestion
        num_modes: The number of modes of this channel.
```

File: mrmustard/lab_dev/circuit_components_utils.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        s: s-parameter of this channel.
```
Comment (SamFerracin) [inline mrmustard/lab_dev/circuit_components_utils.py:39]:
```suggestion
        s: The `s` parameter of this channel.
```

File: mrmustard/lab_dev/circuit_components_utils.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        self.s = s
+
+    @property
+    def representation(self) -> Bargmann:
+        return Bargmann(*triples.displacement_map_s_parametrized_Abc(self.s, len(self.modes)))
+
+
+class CftMap(CircuitComponent):
+    r"""The Complex Fourier Transformation as a channel.
+    This will be used as an internal Channel for representation transformation.
```
Comment (SamFerracin) [inline mrmustard/lab_dev/circuit_components_utils.py:63]:
```suggestion
    
    Used internally as a ``Channel`` for transformations between representations.
```

File: mrmustard/lab_dev/circuit_components_utils.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        )
+        self.s = s
+
+    @property
+    def representation(self) -> Bargmann:
+        return Bargmann(*triples.displacement_map_s_parametrized_Abc(self.s, len(self.modes)))
+
+
+class CftMap(CircuitComponent):
+    r"""The Complex Fourier Transformation as a channel.
```
Comment (SamFerracin) [inline mrmustard/lab_dev/circuit_components_utils.py:62]:
```suggestion
    r"""The Complex Fourier Transformation as a ``Channel``.
```

File: mrmustard/lab_dev/circuit_components_utils.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+    This will be used as an internal Channel for representation transformation.
+
+    Args:
+        num_modes: number of modes of this channel.
```
Comment (SamFerracin) [inline mrmustard/lab_dev/circuit_components_utils.py:66]:
```suggestion
        num_modes: The number of modes of this channel.
```

File: mrmustard/lab_dev/states/base.py [hunk 10]
```diff
@@ -248,11 +253,34 @@ def fock_array(self, shape: Optional[Union[int, Sequence[int]]] = None) -> Compl
         """
         return to_fock(self.representation, shape).array
 
-    def phase_space(self) -> tuple[ComplexMatrix, ComplexVector]:
+    def phase_space(self, s: float, characteristic: bool) -> tuple[ComplexMatrix, ComplexVector]:
         r"""
-        The covariance matrix and the vector of means that describe this state in phase space.
+        The s-parametrized information about the state that describe this state in phase space.
```
Comment (SamFerracin) [inline mrmustard/lab_dev/states/base.py:258]: Can you leave a bit more comments about this?
I am thinking about a user (e.g., me) that sees this for the first time and reads that this gives "s-parametrized information." They'll certainly wonder what this information means.

File: mrmustard/lab_dev/states/base.py [hunk 11]
```diff
@@ -248,11 +253,34 @@ def fock_array(self, shape: Optional[Union[int, Sequence[int]]] = None) -> Compl
...
+
+        Args:
+            s: The parameter of the phase space, which corresponds to the measure of the displacement gate :math:`D_s(\gamma) = e^{\frac{s}{2}|\gamma|^2}D(\gamma)`. For example, :math:`s=0`, the information is related to the Wigner distribution and we have the covariance matrix and means vector of the state.
+            characteristic: whether the phase space is related to the characteristic or not.
```
Comment (SamFerracin) [inline mrmustard/lab_dev/states/base.py:262]:
```suggestion
            characteristic: Whether the phase space is related to the characteristic or not.
```
"Related" how?

Comment (SamFerracin) [inline mrmustard/lab_dev/states/base.py:262]: add `Returns`

File: mrmustard/physics/ansatze.py [hunk 12]
```diff
@@ -530,3 +530,51 @@ def conj(self):
         The conjugate of this ansatz.
         """
         return self.__class__(math.conj(self.array))
+
+
+def Abc_to_cov_and_mean(A: Matrix, b: Vector, c: Scalar) -> Union[Matrix, Vector, Scalar]:
+    r"""Function to transform for ABC Ansatz from quadratic form into the Gaussian form.
```
Comment (SamFerracin) [inline mrmustard/physics/ansatze.py:536]:
```suggestion
    r"""Transforms for ABC Ansatz from quadratic form into the Gaussian form.
```
Also below

File: mrmustard/physics/ansatze.py [hunk 13]
```diff
@@ -530,3 +530,51 @@ def conj(self):
...
+
+    Quadratic form is defined :math:`coeffs * exp(-0.5*(x-mean)^T cov^-1 (x-mean))`.
+    Gaussian form is defined :math:`c * exp(-x^T A x + x^T b)`.
```
Comment (SamFerracin) [inline mrmustard/physics/ansatze.py:539]: Don't forget `Args:` and `Returns:`

File: mrmustard/physics/ansatze.py [hunk 14]
```diff
@@ -530,3 +530,51 @@ def conj(self):
         The conjugate of this ansatz.
         """
         return self.__class__(math.conj(self.array))
+
+
+def Abc_to_cov_and_mean(A: Matrix, b: Vector, c: Scalar) -> Union[Matrix, Vector, Scalar]:
```
Comment (SamFerracin) [inline mrmustard/physics/ansatze.py:535]: I would put all these functions in `triples.py` or in `converters.py`. IMO, they do not belong to `ansatze.py`

**Review by SamFerracin — COMMENTED** [review]
Good job! :100: 

The implementation (math aside) is perfect. I just struggle to follow what's going on, and would appreciate a bit more context in the docs if possible

File: mrmustard/physics/ansatze.py (hunk 14 above)
Comment (sylviemonet) [reply mrmustard/physics/ansatze.py]: I delete this function to open an new PR!

File: mrmustard/lab_dev/states/base.py (hunk 10 above)
Comment (sylviemonet) [reply mrmustard/lab_dev/states/base.py:257]: @ziofil Could you have a look at this doc? I'm struggling in explaining this a little bit more clear in a way....

File: mrmustard/lab_dev/states/base.py [hunk 15]
```diff
@@ -248,11 +252,36 @@ def fock_array(self, shape: Optional[Union[int, Sequence[int]]] = None) -> Compl
         """
         return to_fock(self.representation, shape).array
 
-    def phase_space(self) -> tuple[ComplexMatrix, ComplexVector]:
+    def phase_space(
+        self, s: float, characteristic: bool
+    ) -> tuple[complex, ComplexMatrix, ComplexVector]:
         r"""
-        The covariance matrix and the vector of means that describe this state in phase space.
+        The s-parametrized information about the state that describe this state in phase space.
```
Comment (ziofil) [inline mrmustard/lab_dev/states/base.py:259]:
```suggestion
        Returns the phase space parametrization of a state, consisting in a covariance matrix, a vector of means and a scaling coefficient. When a state is a linear superposition of Gaussians each of cov, means, coeff are arranged in a batch.
        Phase space representations are labelled by an ``s`` parameter (float), which typically takes the values of -1, 0, 1 to indicate Glauber/Wigner/Husimi functions. Note that the same ``(cov, means, coeff)`` triple can be used to parametrize the characteristic functions as well.
```

File: mrmustard/lab_dev/states/base.py [hunk 16]
```diff
@@ -248,11 +252,36 @@ def fock_array(self, shape: Optional[Union[int, Sequence[int]]] = None) -> Compl
...
+
+        Args:
+            s: The parameter of the phase space, which corresponds to the measure of the displacement gate :math:`D_s(\gamma) = e^{\frac{s}{2}|\gamma|^2}D(\gamma)`. For example, :math:`s=0`, the information is related to the Wigner distribution and we have the covariance matrix and means vector of the state.
+            characteristic: Whether the phase space is in the characteristic form or not.
+
+            Returns:
+                The coefficient, the covariance matrix and the mean vector of the state in s-parametrized phase space.
```
Comment (ziofil) [inline mrmustard/lab_dev/states/base.py:266]:
```suggestion
        Args:
            s: The phase space parameter 

            Returns:
                The covariance matrix, the mean vector and the coefficient of the state in s-parametrized phase space.
```

File: mrmustard/lab_dev/states/base.py [hunk 17]
```diff
@@ -248,11 +252,36 @@ def fock_array(self, shape: Optional[Union[int, Sequence[int]]] = None) -> Compl
         """
         return to_fock(self.representation, shape).array
 
-    def phase_space(self) -> tuple[ComplexMatrix, ComplexVector]:
+    def phase_space(
+        self, s: float, characteristic: bool
+    ) -> tuple[complex, ComplexMatrix, ComplexVector]:
```
Comment (ziofil) [inline mrmustard/lab_dev/states/base.py:257]:
```suggestion
    def phase_space(
        self, s: float
    ) -> tuple[ComplexMatrix, ComplexVector, complex]:
```

Comment (ziofil) [reply mrmustard/lab_dev/states/base.py:257]: if you need, modify the order which which the three outputs are returned (everywhere else we leave the coeff last not first)

File: mrmustard/lab_dev/states/base.py [hunk 18]
```diff
@@ -248,11 +252,36 @@ def fock_array(self, shape: Optional[Union[int, Sequence[int]]] = None) -> Compl
...
+                    new_state.representation.ansatz.A[i],
+                    new_state.representation.ansatz.b[i],
+                    new_state.representation.ansatz.c[i],
+                )
+                for i in np.arange(new_state.representation.ansatz.A.shape[0])
+            ]
+
+        raise NotImplementedError(
+            f"Can not calculate phase space for chracteristic function is ``{characteristic}``."
+        )
```
Comment (ziofil) [inline mrmustard/lab_dev/states/base.py:284]:
```suggestion
        return [
            Abc_to_cov_mean_for_state_in_characteristic(
                new_state.representation.ansatz.A[i],
                new_state.representation.ansatz.b[i],
                new_state.representation.ansatz.c[i],
            )
            for i in np.arange(new_state.representation.ansatz.A.shape[0])
        ]
```

Comment (ziofil) [reply mrmustard/lab_dev/states/base.py:284]: ideally `Abc_to_cov_mean_for_state_in_characteristic` would take the batched `A,b,c`. Also, let's find a shorter name for that function.

File: mrmustard/physics/triples.py [hunk 19]
```diff
@@ -459,28 +459,41 @@ def fock_damping_Abc(n_modes: int) -> Union[Matrix, Vector, Scalar]:
...
+
+    The indices of the final triple correspond to the variables :math:`(\gamma_1^*, \gamma_2^*, ..., z_1, z_2, ..., \gamma_1, \gamma_2, ..., z_1^*, z_2^*, ...)` of the Bargmann function of the s-parametrized displacement map, and correspond to ``out_bra, in_bra, out_ket, in_ket`` wires.
 
     Args:
         s: the parametrization related to the ordering of creation and annihilation operators in the expression of any operator. :math:`s=0` is the "symmetric" ordering, which is symmetric under the exchange of creation and annihilation operators, :math:`s=-1` is the "normal" ordering, where all the creation operators are on the left and all the annihilation operators are on the right, and :math:`s=1` is the "anti-normal" ordering, which is the vice versa of the normal ordering. By using s-parametrized displacement map to generate the s-parametrized characteristic function :math:`\chi_s = Tr[\rho D_s]`, and then by doing the complex fourier transform, we get the s-parametrized quasi-probaility distribution: :math:`s=0` is the Wigner distribution, :math:`s=-1` is the Husimi Q distribution, and :math:`s=1` is the Glauber P distribution.
```
Comment (ziofil) [inline mrmustard/physics/triples.py:472]: Move the long explanation to the top and leave a minimal 1-line explanation here

File: mrmustard/physics/ansatze.py [hunk 20]
```diff
@@ -530,3 +530,23 @@ def conj(self):
...
+    Returns:
+        The coefficient, covariance matric and mean vector of the state in phase space.
+    """
+    num_modes = A.shape[-1] // 2
+    Omega = math.J(num_modes).T
+    W = math.conj(math.rotmat(num_modes)).T
+    coeff = c
+    cov = -Omega @ W @ A @ W.T @ Omega.T * settings.HBAR
+    mean = 1j * math.matvec(Omega @ W, b) * math.sqrt(settings.HBAR)
+    return coeff, cov, mean
```
Comment (ziofil) [inline mrmustard/physics/ansatze.py:552]: move `coeff` to last

File: tests/test_lab_dev/test_states.py [hunk 21]
```diff
@@ -261,10 +265,11 @@ def test_to_from_fock(self, modes):
         assert state_in_fock == state_out
 
     def test_to_from_phase_space(self):
-        state0 = Coherent([0], x=1, y=2) >> Attenuator([0], 0.8)
-
-        with pytest.raises(NotImplementedError):
-            state0.phase_space()
+        state0 = Coherent([0], x=1, y=2) >> Attenuator([0], 1.0)
+        coeff, cov, means = state0.phase_space(s=0, characteristic=True)[0]  # batch = 1
```
Comment (ziofil) [inline tests/test_lab_dev/test_states.py:269]: no need for `characteristic`

File: tests/test_lab_dev/test_states.py [hunk 22]
```diff
@@ -100,8 +100,12 @@ def test_from_fock_error(self):
 
     @pytest.mark.parametrize("modes", [[0], [0, 1], [3, 19, 2]])
     def test_to_from_phase_space(self, modes):
-        with pytest.raises(NotImplementedError):
-            Coherent(modes, x=1, y=2).phase_space()
+        coeff, cov, means = Coherent([0], x=1, y=2).phase_space(s=0, characteristic=True)[
```
Comment (ziofil) [inline tests/test_lab_dev/test_states.py:103]: no need for `characteristic`

File: .github/CHANGELOG.md [hunk 23]
```diff
@@ -1,12 +1,19 @@
 # Current develop
 
 ### New features
-  
+* Added one new internal CircuitComponent ``DsMap`` to convert the representations.
```
Comment (ziofil) [inline .github/CHANGELOG.md:4]:
```suggestion
* Added one new private CircuitComponent ``_DsMap`` to convert the representations. 
```

File: mrmustard/physics/ansatze.py [hunk 24]
```diff
@@ -530,3 +530,36 @@ def conj(self):
         The conjugate of this ansatz.
         """
         return self.__class__(math.conj(self.array))
+
+
+def Abc_to_cov_mean_for_state_in_characteristic(
```
Comment (ziofil) [inline mrmustard/physics/ansatze.py:535]:
```suggestion
def Bargmann_Abc_to_phasespace_cov_means(
```

File: mrmustard/lab_dev/states/base.py [hunk 25]
```diff
@@ -43,6 +43,10 @@
 from mrmustard.physics.converters import to_fock
 from mrmustard.physics.gaussian import purity
 from mrmustard.physics.representations import Bargmann, Fock
+from mrmustard.physics.ansatze import (
+    Abc_to_cov_mean_for_state_in_characteristic,
```
Comment (ziofil) [inline mrmustard/lab_dev/states/base.py:47]:
```suggestion
    bargmann_Abc_to_phasespace_cov_means,
```

File: tests/test_physics/test_ansatz.py [hunk 26]
```diff
@@ -20,7 +20,14 @@
 import pytest
 
 from mrmustard import math
-from mrmustard.physics.ansatze import PolyExpAnsatz, ArrayAnsatz
+from mrmustard.physics.ansatze import (
+    PolyExpAnsatz,
+    ArrayAnsatz,
+    Abc_to_cov_mean_for_state_in_characteristic,
```
Comment (ziofil) [inline tests/test_physics/test_ansatz.py:26]:
```suggestion
    bargmann_Abc_to_phasespace_cov_means,
```

File: tests/test_physics/test_ansatz.py [hunk 27]
```diff
@@ -284,3 +291,54 @@ def test_algebra_with_different_shape_of_array_raise_errors(self):
 
         with pytest.raises(Exception):
             aa1 == aa2
+
+    def test_Abc_to_cov_mean_for_state_in_characteristic(self):
```
Comment (ziofil) [inline tests/test_physics/test_ansatz.py:295]:
```suggestion
    def test_bargmann_Abc_to_phasespace_cov_means(self):
```

File: tests/test_physics/test_ansatz.py [hunk 28]
```diff
@@ -284,3 +291,54 @@ def test_algebra_with_different_shape_of_array_raise_errors(self):
...
+        state_cov = np.array([[0.32210229, -0.99732956], [-0.99732956, 6.1926484]])
+        state_means = np.array([0.2, 0.3])
+        state = DM.from_bargmann([0], wigner_to_bargmann_rho(state_cov, state_means))
+        state_after = state >> _DsMap(modes=[0], s=0)  # pylint: disable=protected-access
+        A1, b1, c1 = state_after.bargmann_triple
+        (
+            new_state_cov,
+            new_state_means,
+            new_state_coeff,
+        ) = Abc_to_cov_mean_for_state_in_characteristic(A1, b1, c1)
```
Comment (ziofil) [inline tests/test_physics/test_ansatz.py:306]:
```suggestion
        ) = bargmann_Abc_to_phasespace_cov_means(A1, b1, c1)
```

File: tests/test_physics/test_ansatz.py [hunk 29]
```diff
@@ -284,3 +291,54 @@ def test_algebra_with_different_shape_of_array_raise_errors(self):
...
+        A, b, c = wigner_to_bargmann_rho(state_cov, state_means)
+        state = DM.from_bargmann(modes=[0, 1], triple=(A, b, c))
+
+        state_after = state >> _DsMap(modes=[0, 1], s=0)  # pylint: disable=protected-access
+        A1, b1, c1 = state_after.bargmann_triple
+        (
+            new_state_cov1,
+            new_state_means1,
+            new_state_coeff1,
+        ) = Abc_to_cov_mean_for_state_in_characteristic(A1, b1, c1)
```
Comment (ziofil) [inline tests/test_physics/test_ansatz.py:329]:
```suggestion
        ) = bargmann_Abc_to_phasespace_cov_means(A1, b1, c1)
```

**Review by ziofil — APPROVED** [review]
good to go!

File: mrmustard/lab_dev/states/base.py [hunk 30]
```diff
@@ -248,11 +252,27 @@ def fock_array(self, shape: Optional[Union[int, Sequence[int]]] = None) -> Compl
...
+
+            Returns:
+                The covariance matrix, the mean vector and the coefficient of the state in s-parametrized phase space.
         """
-        raise NotImplementedError
+        if not isinstance(self.representation, Bargmann):
+            raise ValueError(f"Can not calculate phase space for ``{self.name}`` object.")
+
+        new_state = self >> _DsMap(self.modes, s=s)  # pylint: disable=protected-access
+        return Abc_to_cov_mean_for_state_in_characteristic(
```
Comment (sylviemonet) [inline mrmustard/lab_dev/states/base.py:271]:
```suggestion
        return bargmann_Abc_to_phasespace_cov_means(
```

**Pull request merged**